SPECIFICATION.md

```diff
@@ -200,3 +200,72 @@ Whenever a new feature is added to the SDK that does not induce a breaking chang
 ## Patch Versions
 
 Whenever a new release is warranted to only preform bug fixes it will be considered a patch version.
+
+# Git Structure
+
+As the APIs evolve over time, there will be instances were multiple versions of the SDKs will need to be supported at the same time. To better facilitate this, the Server SDK repositories will follow a common version branching structure for releases. This structure is similiar to, but not exactly, [Gitflow](https://nvie.com/posts/a-successful-git-branching-model/).
+
+## Branch Structure
+
+The SDKs will follow a `MAJOR.MINOR` release branch naming structure, with mainline work being done in a `MAJOR.x` branch. `MAJOR.x` will be assigned as the default branch for the SDK. `master` will no longer be used as a branch name at all. 
+
+When a new MINOR release is to be released, `MAJOR.x` will be branched to form `MAJOR.MINOR`, and the point release can be tagged off of the appropriate minor branch. For example, `3.0.0` would be inside the `3.0` branch, and exist as a tag. `3.0.1` would be a tag inside the `3.0` branch as well.
+
+**Branch Example:**
+Given an SDK that is currently working against a `4.x` release, the branch structure would look like the following:
+
+* `4.x`
+* `1.0`
+* `1.1`
+* `1.2`
+* `2.0`
+* `3.0`
+* `3.1`
+* `3.2`
+* `3.3`
+* `4.0`
+* `4.1`
+
+## Workflows
+
+Most work will be for one of two things - new features or bug fixes. The only major workflow differences are determining which branch that the work should be based on. Major releases will function as New Features, but have a few additional steps needed to make the appropriate branches.
+
+### New Features
+
+New features are always based on, and pull requested against, the current default branch. For example, if the SDK is using `4.x` as the default branch, the new feature will be based on that branch. This is the same workflow as if we had a `master` branch. The pull request will be reviewed and will be available with the next minor version release. 
+
+New features will not be made available to previous MAJOR releases, even if the MAJOR release is still in a support window. For example, if the current MAJOR version is 4.x, and 3.x is still supported, a new API feature will be created against and only be available in the 4.x line.
+
+### Bug Fixes
+
+For bug fixes, a pull request and patch should be made against the oldest supported point release that the bug exists in. This allows the fix to be up-merged to all releases that contain the bug. For example, say we currently support the 3.x and 4.x lines of the SDK, and a new bug is discovered in 3.9.0, but after triaging was actually introduced in 3.5.0. The PR would be based on the `3.5` branch, and the pull request set to merge back into `3.5`. Once the pull request is approved and merged into `3.5`, `3.5` can then be merged into any newer supported branches, extending the bugfix to any supported versions.
```

Comment:
> but if we're abiding by semver we should only be maintaining 1 release per Major version still under support

Semver only dictates how your version numbers work, it doesn't dictate anything about support. We can decide how granular we want to be when it comes to support. Semver doesn't care if we maintain 1, 2, or 12 branches at various revision numbers, it just tells us how to determine the next version number based on the code we are maintaining.

> we should expect our users to upgrade to the latest minor of a given major.

Many don't, as given by our usage stats. We should determine in greater detail if users are even upgrading within MINOR releases (not that we tend to have a bunch).

> Otherwise, if our SLA is say 6 months for example, and I've released 8 minor versions in the last year, and the bug affects the oldest minor version, I'd have to cut 8 releases for a single bug fix - this doesn't make sense to me.

No, just any releases made in the last 6 months still covered by the SLA. So, let's say we had the following releases:

* 3.6.0 - January
* 3.6.1 - January
* 3.7.0 - February
* 3.6.2 - March
* 3.7.1 - March
* 4.0.0 - June
* 4.1 - August
* 3.7.2 - August
* 3.8.0 - September
* 4.2 - November

And it's currently December 2nd, we would only care about the MINORs and MAJORs released from July through now: 4.1, 3.7, 3.8, and 4.2. If a bug is found in 3.6.1, we still wouldn't fix anything earlier than those minor branches from August. You would check if it exists in the 3.7 release line, and if so fix it there and migrate it up.

Reply:
I'd agree with Steve here. Backporting fixes to major versions seems like more work than we need to do.

As there are no breaking changes in minor versions, why wouldn't they upgrade from 3.5 to 3.9 to get the fix if they need it?

Are you aiming for those that do update dependencies, but are pinned to minor versions?

Reply:
I changed the wording to just have `MAJOR.0` branches. This will cut down on the amount of ports we need to do.

> As there are no breaking changes in minor versions, why wouldn't they upgrade from 3.5 to 3.9 to get the fix if they need it?
> 
> Are you aiming for those that do update dependencies, but are pinned to minor versions?

Exactly. We don't seem to see users even updating inside minor release paths, and staying at patch releases.